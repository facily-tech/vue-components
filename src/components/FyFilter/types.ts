export interface IFiltersAllTags {
  key: string | number;
  value: string;
  enable: boolean;
  type: string;
}
