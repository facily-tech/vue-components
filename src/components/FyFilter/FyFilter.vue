<template>
  <v-card class="filter-wrapper" elevation="0">
    <v-card-text v-if="checkAllTags()">
      <v-row dense>
        <v-col cols="12">
          <ul class="d-flex no-wrap pa-0" style="list-style: none">
            <li v-for="(item, key) in allTags" :key="key">
              <v-chip
                v-if="item.enable"
                class="ml-0 ma-2"
                close
                color="primary"
                text-color="white"
                @click:close="$emit('remove-tag', item)"
              >
                {{ item.value }}
              </v-chip>
            </li>
          </ul>
        </v-col>
      </v-row>
    </v-card-text>
    <v-card-text>
      <v-row dense>
        <v-col md="9" class="text-left">
          <slot name="content"></slot>
        </v-col>
        <v-col md="3" class="text-right">
          <fy-button
            v-if="checkAllTags()"
            outlined
            color="primary"
            dark
            @click="$emit('clear-tags')"
            data-test="btn-clear-filter"
          >
            <v-icon> mdi-filter-off </v-icon>
            <span style="text-transform: capitalize"> Limpar</span>
          </fy-button>
          <fy-button
            color="primary"
            outlined
            class="ml-5"
            :disabled="!validForm"
            @click="$emit('click-filter')"
            data-test="btn-filter"
          >
            <v-icon>mdi-filter</v-icon>
            <span style="text-transform: capitalize"> Filtrar </span>
          </fy-button>
        </v-col>
      </v-row>
    </v-card-text>
  </v-card>
</template>

<script lang="ts">
import Vue from 'vue';

import { FyButton } from '@/index';

import { IFiltersAllTags } from './types';

export default Vue.extend({
  name: 'fy-filter',

  components: { FyButton },

  props: {
    allTags: {
      type: Array as () => IFiltersAllTags[],
      default: [] as IFiltersAllTags[],
      required: true,
    },
    validForm: { type: Boolean, default: false, required: true },
  },

  data() {
    return {
      count: 0,
    };
  },

  watch: {
    allTags: {
      handler() {
        this.checkFilters();
      },
      deep: true,
    },
  },

  methods: {
    checkAllTags(): boolean {
      return this.allTags?.some((item) => item.enable);
    },

    checkFilters(): void {
      this.count = 0;
      this.allTags.map((item) => {
        if (item.enable) this.count++;
      });
    },
  },
});
</script>
