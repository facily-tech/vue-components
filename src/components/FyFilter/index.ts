export { default as FyFilter } from './FyFilter.vue';
