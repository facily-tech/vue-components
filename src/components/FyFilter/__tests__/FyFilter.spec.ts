import Vue from 'vue';
import Vuetify from 'vuetify';
import FyFilter from '../FyFilter.vue';
import { shallowMount, Wrapper } from '@vue/test-utils';

Vue.use(Vuetify);

describe('FyFilter.ts', () => {
	let WRAPPER: Wrapper<Vue>;
	beforeEach(() => {
		WRAPPER = shallowMount(FyFilter, {
			vuetify: new Vuetify(),
			propsData: {
				count: 0,
				allTags: [
					{
						key: 0,
						value: 'Nome',
						enable: true,
						type: 'q',
					},
				],
				validForm: true,
			},
		});
	});

	test('Component exists', () => {
		expect(WRAPPER.exists()).toBeTruthy();
	});

	test('checkAllTags test', async () => {
		WRAPPER.vm.checkAllTags();
		await WRAPPER.vm.$nextTick();
		expect(WRAPPER.vm.allTags[0].enable).toBeTruthy();
	});

	test('checkFilters test', async () => {
		WRAPPER.vm.count = 0;
		WRAPPER.vm.checkFilters();
		await WRAPPER.vm.$nextTick();
		expect(WRAPPER.vm.count).toBe(1);
	});

	test('checkFilters test else block', async () => {
		WRAPPER.vm.count = 0;
		WRAPPER.vm.allTags = [
			{
				key: 0,
				value: 'Nome',
				enable: false,
				type: 'q',
			},
		];
		WRAPPER.vm.checkFilters();
		await WRAPPER.vm.$nextTick();
		expect(WRAPPER.vm.count).toBe(0);
	});
});
