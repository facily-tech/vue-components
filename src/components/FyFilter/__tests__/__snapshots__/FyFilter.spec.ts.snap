// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FyFilter.ts should render component and match snapshot 1`] = `
<div class="filter-wrapper v-card v-sheet theme--light elevation-0">
  <!---->
  <div class="v-card__text">
    <div class="row row--dense">
      <div class="text-left col-md-9 col"></div>
      <div class="text-right col-md-3 col">
        <!----> <button type="button" disabled="disabled" class="ml-5 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default" data-test="btn-filter" color="primary" outlined=""><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-filter theme--light"></i> <span style="text-transform: capitalize;"> Filtrar </span></span></button>
      </div>
    </div>
  </div>
</div>
`;
